### ISSUE TITLE
Net::Amazon::S3::Client bucket object wrongly claims corruption and fails

### ISSUE THREAD

**Issue by rnavarr** [pfig/net-amazon-s3#63]
Folks,
I am using Net::Amazon::S3::Client v0.60 with Perl 5.16 and the program:
```perl
#!/usr/bin/env perl

use strict;
use warnings;

use Net::Amazon::S3;
use Net::Amazon::S3::Client;

use lib '../lib';
use BMaws;  # for credential fetch

my $s3fname = q{s3://boostde/dataprep/searchq/searchq20150401053400.19177/adwords/3090/3830538210_adwords_searchq20150401053400.19177.csv.gz};

my ($ac, $sac) = get_aws_creds();
my $s3 = Net::Amazon::S3->new(
     aws_access_key_id=> $ac,
     aws_secret_access_key=> $sac,
     retry=>1,
     );
my $client = Net::Amazon::S3::Client->new( s3 => $s3 );
my ($base_bucket) = $s3fname =~ m!s3://([^/]+)/!;

my $bucket = $client->bucket( name => $base_bucket );
my ($s3key) = $s3fname =~ s!^s3://[^/]+/!!r;
my $object = $bucket->object( key => $s3key );

print STDERR $object->get;
```
which accesses an S3 object with the properties (from AWS console):
```
Bucket: 	boostde
Folder: 	3090
Name: 	3830538210_adwords_searchq20150401053400.19177.csv.gz
Link: 	
[This is a private link] https://s3-us-west-2.amazonaws.com/boostde/dataprep/searchq/searchq20150401053400.19177/adwords/3090/3830538210_adwords_searchq20150401053400.19177.csv.gz
Size: 	25578798
Last Modified: 	Tue Mar 31 23:03:46 GMT-700 2015
Owner: 	boostctr
ETag: 	d42d103edb0d482ba25934f0395b7ed4-5
Expiry Date: 	None
Expiration Rule: 	N/A
```
the S3 data is fine as I can show with the aws command line tools:
```
aws s3 cp s3://boostde/dataprep/searchq/searchq20150401053400.19177/adwords/3090/3830538210_adwords_searchq20150401053400.19177.csv.gz - | gunzip | tail -1
Total, --, --, --, --, --,193128,16999385,2.5, --, --, --,1739814,0.00%,0.00%,0.0,0,0,336006960000,0,0, --,8.80%, --, --, --, --,19765830, --, --, --, --,0.0,0.0,0
```
Running the program fails with:
```
% perl bugs3lib.pl
Corrupted download at /usr/local/share/perl5/Net/Amazon/S3/Client/Object.pm line 103.
        Net::Amazon::S3::Client::Object::_get('Net::Amazon::S3::Client::Object=HASH(0x3be81d0)') called at /usr/local/share/perl5/Net/Amazon/S3/Client/Object.pm line 111
        Net::Amazon::S3::Client::Object::get('Net::Amazon::S3::Client::Object=HASH(0x3be81d0)') called at bugs3lib.pl line 27
```
Checking /usr/local/share/perl5/Net/Amazon/S3/Client/Object.pm I can see that the multi-part Etag is not matching the md5_sum of the actual data. No idea which tool used multi-part upload. The point is this Perl library simply fails to get that valid object. 

I tweaked the library by avoiding the md5_sum check when confronted with a multi-part Etag. This makes the program work again...but of course I see no way to check for download corruption in the case when S3 is storing only a multi-part Etag. 

My diff:
```
% diff Object.pm.orig Object.pm.patch
104c104
<             if $self->_etag($http_response) ne $md5_hex;
---
>             if (!$self->_is_multipart_etag($etag) && $self->_etag($http_response) ne $md5_hex);
```
Ideally AWS needs to store the actual MD5 sum of the data as well as what it does (or allow the actual MD5 sum to be derived from it).

Any suggestions for me that do not involve tweaking the library?

Cheers,

R

**Comment by rustyconover** [comment]
Since the multipart upload parts can be of any length, it is not possible to calculate the multipart ETag after it is created without knowing the length of the individual parts.  If you need to insist on ETag's being correct rather than multipart in nature, you could issue a COPY request on the key (to the same key) after it has been uploaded on S3 which should reset the ETag to be the actual MD5 checksum of the data once it is complete regardless if multipart upload was used.

Cheers,

Rusty

**Comment by rnavarr** [comment]
Hi Rusty thanks for the suggestion. 
Is there something wrong with my reasoning that the library has a bug? 
I certainly have no knowledge of who sends or how files make their way to S3, one would like Perl to not fail wrongly in any case.
In the mean time I am making progress having turned off the failing corruption check...

**Comment by lackas** [comment]
Hi @rnavarr, yes this was a bug (should not have checked the checksum for multi-part uploads). Is fixed in the latest version 0.80, though.

**Comment by pfig** [comment]
Hi,

Please update to 0.80 and if this still persists please open an issue in @rustyconover's fork, https://github.com/rustyconover/net-amazon-s3.
